packages/react-dom-bindings/src/client/ReactFiberConfigDOM.js

```diff
@@ -1283,6 +1291,26 @@ export function registerSuspenseInstanceRetry(
   instance._reactRetry = callback;
 }
 
+export function canHydrateFormStateMarker(
+  instance: HydratableInstance,
+): null | FormStateMarkerInstance {
+  const nodeData = (instance: any).data;
+  if (
+    nodeData === FORM_STATE_IS_MATCHING ||
+    nodeData === FORM_STATE_IS_NOT_MATCHING
+  ) {
+    const markerInstance: FormStateMarkerInstance = (instance: any);
+    return markerInstance;
+  }
+  return null;
+}
```

Comment:
There is this unfortunate forking of behavior in the `canHydrate...` functions where if we are in a Singleton like `<body>` or are in the container scope we assume there may be 3rd party nodes we need to skip over. If our `nextHydratable` was some injected 3rd party stylesheet then hydration would fail as this is currently implemented. I think this function needs. a similar treatment as `canHydrateSuspenseInstance`

This will make your optimization where we omit the marker when the hook is normal harder too b/c it'll need to be able to scan forward but also jump back to where it was if it doesn't find anything

Reply:
Re: the last point about the optimization, the idea is we wouldn't look for these nodes at all in that case. Basically it would revert to how it is today.

Reply:
hmm I think this would still be a problem

```
function App() {
  ... = useFormState();
  return <div>...</div>
}


<script injectedBy3rdParty />
<!--!F-->
<div>
  ...
</div>
```
If I'm hydrating App the nextHydratableInstance will end up on `<script injectedBy3rdParty />` Then when we check whether there is a form marker we see there isn't and we assume no postback state. Then we try to claim the next hydratable for `<div>`. We end up hitting the form state marker and stop causing a hydration mismatch.

Reply:
yeah that part makes sense, I just in the case where there's no postback state, there are no extra comment nodes at all, so all this code gets skipped anyway